packages/fx-core/src/plugins/resource/bot/v2/azureHosting.ts

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Inputs, ResourceTemplate, Void } from "@microsoft/teamsfx-api";
+import { BicepTemplate, Context } from "@microsoft/teamsfx-api/build/v2";
+import * as fs from "fs-extra";
+import path from "path";
+import { BotHostTypes, generateBicepFromFile } from "../../../../common";
+import { ArmTemplateResult } from "../../../../common/armInterface";
+import { Bicep } from "../../../../common/constants";
+import { getTemplatesFolder } from "../../../../folder";
+import { getActivatedV2ResourcePlugins } from "../../../solution/fx-solution/ResourcePluginContainer";
+import { NamedArmResourcePluginAdaptor } from "../../../solution/fx-solution/v2/adaptor";
+import { HostTypes, PluginBot } from "../resources/strings";
+import { BicepConfigs } from "./botSolution";
+import * as utils from "../utils/common";
+
+export interface BicepFiles {
+  provisionOrchestrationFile: string;
+  provisionModuleFile: string;
+  configOrchestrationFile?: string;
+  configModuleFile?: string;
+}
+
+export interface BicepModules {
+  provisionOrchestration: string;
+  provisionModule: string;
+  configOrchestration?: string;
+  configModule?: string;
+}
+
+async function generateBicep(
+  bicepFiles: BicepFiles,
+  generate: (filename: string) => Promise<string>
+): Promise<BicepModules> {
+  let filenames = [bicepFiles.provisionOrchestrationFile, bicepFiles.provisionModuleFile];
+  if (bicepFiles.configModuleFile && bicepFiles.configOrchestrationFile) {
+    filenames = filenames.concat([bicepFiles.configOrchestrationFile, bicepFiles.configModuleFile]);
+  }
+
+  const modules = await Promise.all(filenames.map((filename) => generate(filename)));
+
+  return {
+    provisionOrchestration: modules[0],
+    provisionModule: modules[1],
+    configOrchestration: modules?.[2],
+    configModule: modules?.[3],
+  };
+}
+
+export abstract class AzureHosting {
+  abstract bicepFolderRelativeDir: string;
+  abstract hostType: string;
+  abstract configurable: boolean;
+
+  reference: any = undefined;
+  pluginId: string;
+  constructor(pluginId: string) {
```

Comment:
If this pluginId is only used in generateBicep(), I prefer to make it as parameter of that method, so that we can remove it from constructor and make AzureHosting be a stateless class.

There are some benefits about stateless
1. there is no state related issue/bugs
2. it is easy to be shared as utility class